src/Sarif/Readers/LineMappingStreamReader.cs

```diff
@@ -13,123 +12,142 @@ namespace Microsoft.CodeAnalysis.Sarif.Readers
 
-        public override int Read(char[] buffer, int index, int count)
-        {
-            // Track the first line number and char total before the current buffer
-            if (_lineCount > 0)
+            // Find the correct newline
+            for (; current.LineNumber < line && bufferIndex < _bufferLength; bufferIndex++)
             {
-                _firstLineNumber += _lineCount;
-                _firstLineCharsBeforeBuffer = 0;
+                if (_buffer[bufferIndex] == '\n')
+                {
+                    current.ByteOffset = current.ByteOffset + this.CurrentEncoding.GetByteCount(_buffer, current.BufferIndex, bufferIndex - current.BufferIndex);
+
+                    current.LineNumber++;
+                    current.CharInLine = 0;
+                    current.BufferIndex = bufferIndex;
+                }
             }
+           
+            // Count bytes on this line
+            if (current.CharInLine < charInLine)
+            {
+                int charsToAdd = charInLine - current.CharInLine;
+                if (current.BufferIndex + charsToAdd > _bufferLength) throw new ArgumentOutOfRangeException($"Position must be in buffer from last read. ({line}, {charInLine}) requested; ({current.LineNumber}, {current.CharInLine + _bufferLength - current.BufferIndex}) is end of buffer.");
 
-            _firstLineCharsBeforeBuffer += _lastLineChars;
+                current.ByteOffset = current.ByteOffset + this.CurrentEncoding.GetByteCount(_buffer, current.BufferIndex, charsToAdd);
 
-            // Maintain total byte count read
-            _bytesReadPreviously += _bytesRead;
+                current.CharInLine = charInLine;
+                current.BufferIndex += charsToAdd;
+            }
 
-            // Read the new buffer
-            int charsRead = base.Read(buffer, index, count);
+            // Return the found position
+            return current;
+        }
 
-            // Copy buffer so we can map char in line to byte count (real buffer can be shifted by reader, invalidating indices)
-            if (_buffer == null || _buffer.Length < buffer.Length) _buffer = new char[buffer.Length];
-            Buffer.BlockCopy(buffer, 2 * index, _buffer, 0, 2 * charsRead);
-            _bufferIndex = 0;
-            _bufferLength = charsRead;
+        private FilePosition CountUpTo(int index)
+        {
+            if (index < 0 || index >= _bufferLength) throw new IndexOutOfRangeException("index");
 
-            // Ensure space to hold start of each line
-            if (_lineStartByteOffsets == null || _lineStartByteOffsets.Length < charsRead) _lineStartByteOffsets = new int[charsRead];
-            if (_lineStartIndices == null || _lineStartIndices.Length < charsRead) _lineStartIndices = new int[charsRead];
-            _lineCount = 0;
+            FilePosition current = _lastReturnedPosition;
+            if (_lastReturnedPosition.BufferIndex > index) current = _bufferStartPosition;
```

Comment:
>current = _bufferStartPosition; [](start = 59, length = 31)

Put this on its own line with curly braces. I find this improves readability. StyleCop 1503

Reply:
OK so I confirmed that our coding standard permits the single-line format, but it needs to have curly braces.
if (blah) { method(); }

---
In reply to: [222033337](https://github.com/Microsoft/sarif-sdk/pull/1106#discussion_r222033337) [](ancestors = 222033337)